master/buildbot/changes/gitpoller.py

```diff
@@ -338,27 +344,67 @@ def _process_changes(self, newRev, branch):
                 repository=bytes2unicode(self.repourl, encoding=self.encoding),
                 category=self.category, src=u'git')
 
+    def _isSshPrivateKeyNeededForCommand(self, command):
+        commandsThatNeedKey = [
+            'fetch',
+            'ls-remote',
+        ]
+        if self.sshPrivateKey is not None and command in commandsThatNeedKey:
+            return True
+        return False
+
+    def _downloadSshPrivateKey(self, keyPath):
+        with open(keyPath, 'w') as keyFile:
```

Comment:
```
    try:
        oldmask = os.umask(0o077)
        with open(keyPath, 'w') as keyFile:
            keyFile.write(self.sshPrivateKey)
    finally:
        os.umask(oldmask)
```

Reply:
additional merit points for writing an utility context manager in buildbot.utils
..and make sure this works on Windows, hint: tests!

Reply:
I now realize that you stole this code from the worker/transfer code.

That code actually is 12 years old (https://github.com/buildbot/buildbot/commit/05317ec1a0bc96f8c66910235b2de228e173a29a#diff-af6cb5744179a15d0117a8201da0995dR819), so I think it is time to not trust the comment from @warner, and just use the try: finally method.

Reply:
The way I've written the code definitely has a race condition, thus the comment :-) But even the umask method has a race condition: if the attacker creates keyPath just after umask, but before open, the permissions won't be changed and the key file would retain the original permissions. Also, if something on another thread depends on os.umask value, then this would result in a race condition that might affect correctness and would be a hell to debug.

Wouldn't it be a better idea just to rely on the permissions of the parent directory?

Reply:
> if the attacker creates keyPath just after umask, but before open, the permissions won't be changed

that is what O_EXCL is made for. I agree umask is not thread safe and no good solution.
```

import os

flags = os.O_WRONLY | os.O_CREAT | os.EXLC
with os.fdopen(os.open('testfile.txt', flags, 0o600), 'w') as fout:
    fout.write("secrets!")
```
I fear that might be posix only.

> Wouldn't it be a better idea just to rely on the permissions of the parent directory?

Then we would have to check that.

Reply:
What do you think about something like this?
```
os.mkdir('path/to/keydir', mode=0o700)
    with open('path/to/keydir/key, 'w') as keyFile:
        keyFile.write(self.sshPrivateKey)
    <do stuff>
shutil.rmtree('path/to/keydir')```